man/man1/duel.1
.TH duel 1
.SH NAME
\fBduel\fR - Stages a duel using two opposing instances of \fBbishbosh\fR
.SH SYNOPSIS
\fBbishbosh\fR [\fIOPTIONS\fR]
.SH DESCRIPTION
.PP
An application which automatically plays chess, by forking two independently configured instances of \fBbishbosh\fR.
One instance automates the moves of White & the other automates the moves of Black.
This application merely acts as an intermediary, passing moves between these instances to synchronise their positions,
while measuring the total time taken by either side & collating the game-results to permit comparison of the fitness of their respective configurations.
.PP
This process facilitates optimisation of the complex configuration of \fBbishbosh\fR.
.PP
CAVEAT: the population of games resulting from this statistical experiment must be diverse, but this application's logic is largely deterministic (& both players are configured to use it).
In order to introduce some randomness, PGN-databases must be configured in which the application may search for positions matching the current game.
When more than one match for the current position is found, one of them is \fIrandomly\fR selected at the source of the next move.
If the specified PGN-databases fail to provide such move-options, deterministic play will result.
.SH OPTIONS
.TP
\fB--verbosity=\fR(\fBSilent\fR|\fBNormal\fR|\fBVerbose\fR|\fBDeafening\fR), defaulting to "\fBNormal\fR".
Defines the quantity of ancillary output required.
.TP
\fB-d\fR \fIInt\fR, \fB--nDecimalDigits=\fR\fIInt\fR, defaulting to "\fB0\fR".
The precision with which to display the total time taken by each side.
.TP
\fB-n\fR \fIInt\fR, \fB--nGames=\fR\fIInt\fR, defaulting to "\fB1\fR".
The number of games to be played before presenting the results.
.TP
\fB-t\fR \fIInt\fR, \fB--readTimeout=\fR\fIInt\fR, defaulting to \fIindefinitely\fR.
The maximum number of seconds to wait for one of the forked processes to return a move, before aborting.
CAVEAT: the required period typically required depends on;
.IP \(bu
the machine's speed,
.IP \(bu
what else is running on the machine,
.IP \(bu
the size of the machine's RAM,
.IP \(bu
the time initially required to process any PGN-database,
.IP \(bu
\fBbishbosh\fR's configuration (e.g. search-depth or fitness-evaluation criteria).
.TP
.B --verifyConfiguration
Verify the mutual compatibility of the two configuration-files before forwarding each to a forked instance of \fBbishbosh\fR,
where the syntactic & semantic correctness will be independently verified.
The configurations are required to define the same move-notation & to define the logical colour of the automated player in the correct order to avoid deadlock.
N.B. since verification of this compatibility-requirement necessitates parsing, syntactic-verification is an inevitable side-effect.
.TP
\fB-i\fR \fIFile-path\fR, \fB--appendInputConfigFilePath=\fR\fIFile-path\fR; there is no default.
Define the path to a configuration-file to forward to \fBbishbosh\fR.
Two such paths are required, the first will be passed to the instance playing White & the second to the instance playing Black.
.SS "Generic Program-information"
.TP
\fB-v\fR, \fB--version\fR
Outputs version-information & then exits.
.TP
\fB-?\fR, \fB--help\fR
Displays a usage-message & then exits.
.SH EXIT-STATUS
\fB0\fR on success, & >\fB0\fR if an error occurs.
.SH EXAMPLES
.nf
.B duel --nGames=32 --verbosity='Verbose' --verifyConfiguration -i 'config/Raw/bishbosh_duel_white.xml' -i 'config/Raw/bishbosh_duel_black.xml'
.fi
.IP \(bu
The application has been instructed to provide a relatively large amount of feedback.
.IP \(bu
Two packaged configuration-files have been referenced.
.IP \(bu
The compatibility of the configuration-files is verified to confirm that they automate White & Black respectively, & that they use the same move-notation.
.IP \(bu
CAVEAT: assumes that the executable exists on \fB$PATH\fR, otherwise an explicit path will be required.
.PP
.nf
.B make duel
.fi
.IP \(bu
Much the same as the previous example, but verification is unnecessary since the configuration-files are hard-coded.
.IP \(bu
The current working directory must be \fBbishbosh\fR's installation-directory where the required \fBmakefile\fR is defined.
.SH FILES
.TS
lb	lb
l	l
lb	l	.
File-name	Contents
=========	========
config/Raw/bishbosh_duel_{black,white}.xml	Sample configuration-files.
makefile	Facilitates common tasks.
man/man1/bishbosh.1	The man-page for \fBbishbosh\fR.
.TE
.SH AUTHOR
Written by Dr. Alistair Ward.
.SH BUGS
.PP
In the absence of any configured \fBpgnOptions\fR, the game is completely deterministic & will reproduce the previous result endlessly.
Even when PGN-databases have been configured, they may introduce insufficient randomness to produce a large population of unique games on which to perform statistical analysis.
.SS "REPORTING BUGS"
Report bugs to "\fBbishbosh@functionalley.com\fR".
.SH COPYRIGHT
Copyright \(co 2021 Dr. Alistair Ward
.PP
This program is free software: you can redistribute it and/or modify it under the terms of the GNU General Public License as published by the Free Software Foundation, either version 3 of the License, or (at your option) any later version.
.PP
This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License for more details.
.PP
You should have received a copy of the GNU General Public License along with this program. If not, see <\fBhttps://www.gnu.org/licenses/\fR>.
.SH "SEE ALSO"
.IP \(bu
Home-page: <\fBhttps://functionalley.com/BishBosh/bishbosh.html\fR>.
.IP \(bu
<\fBhttps://hackage.haskell.org/package/bishbosh\fR>.
.IP \(bu
<\fBhttps://github.com/functionalley/BishBosh\fR>.
.IP \(bu
Source-documentation is generated by "\fBHaddock\fR", & is available in the distribution.
.IP \(bu
<\fBhttps://www.haskell.org/haddock/\fR>.



changelog.markdown
# 2018-01-01 Dr. Alistair Ward <bishbosh@functionalley.com>

## 0.0.0.1
* First version of the package.

## 0.0.0.2
* Minor changes required to build on Windows.

## 0.0.0.3
* Added **Paths_bishbosh** to **Other-modules** section of cabal file.

## 0.0.0.4
* Changed references to author's domain-name.

## 0.0.0.5
* Added ability to specify the text-encoding used in a PGN-database file.
* Updated list of test-compilers.

## 0.0.0.6
* Fixed failure to persist game-state after requesting that the game be restarted.
* Fixed parsing of **TextEncoding** in **PGNOptions**.
* Replaced module **Distribution.Verbosity** with **BishBosh.Input.Verbosity**.

## 0.0.0.7
* Rewrote function **BishBosh.Data.RoseTree.countTerminalNodes** in accordance with the suggestions of David Feuer.
* Amended function **BishBosh.State.EnPassantAbscissa.mkMaybeEnPassantAbscissa** to guard against exposing one's King after En-passant capture.

## 0.0.0.8
* Corrected the parsing of FEN when an Enpassant-destination defined on file **b** was erroneously interpreted as a bishop in the previous **CastleableRooks** field.
* Added parent class **BishBosh.Property.ExtendedPositionDescription.EPD** for **Property.ForsythEdwards.FEN**, for which the latter typically has a default implementation of both methods.

## 0.1.0.0
### Bug-fixes:
* Upgraded the transposition-table in module **Search.DynamicMoveData**, from merely recording moves (which doesn't include the rank to which a Pawn is promoted), to recording qualified-moves.
* In function **Search.AlphaBeta.negaMax.descend.selectMax**, amended bound function **isFitter** to prefer shorter move-sequences where fitness is equal, & corrected the scenario in which all nodes were skipped because they were repetitious, but without ever defining alpha.

### Features:
* Added the configurable runtime ability to asynchronously decompress PGN-databases, & to set a maximum number of games to read.
* Added a configuration-option to normalise the values of specified piece-square tables into the closed unit-interval.
* Modularised the packaged config-files, by defining XML **External Entities** in the DTD.
* Added suggestions on failure to parse a user-command, & created a module **Text.AutoComplete** to contain common code.
* Added **makefile** to facilitate common tasks.
* Removed the configuration-option **preferMovesTowardsCentre** & its implementation in function **Cartesian.Coordinates.radiusSquared**, because of it's conceptually wobbly foundations.

### New Runtime Cmmands:
-----------------------------------------
Command				| Purpose
--------------------------------| -------
**availableMoves**		| to report all available moves from the current position.
**maxPositionInstances**	| to reveal the maximum number of instances any available position has been visited.
**reversiblePlyCount**		| to count the number of consecutive reversible plies that have been made.
-----------------------------------------

### Command-line Options:
* Added a new module **Input.CategorisedCommandLineOptions** to improved the partitioning of command-line options into functional categories.
* Added a command-line option **--formatPieceSquareTableForGNUPlot** to print the piece-square tables in a format suitable for **GNUPlot**.

### Performance:
* Included a compilation-flag **unboxedarrays**, to request the use of unboxed arrays where (infrequently) possible.
* Changed data-type **Component.PieceSquareByCoordinatesByRank.EitherPieceSquareValueByNPiecesByCoordinates**, bringing type **Cartesian.Coordinates.ByCoordinates** inside **Either**, leading to significant space/time gains.
* Constructed each large constant data-structure in parallel. Bracketed all data-parallel operations with CPP-conditionals controlled by the compilation-flag **threaded**.
* Parallelised function **Attribute.CriterionValue.calculateWeightedMean**.

### New Modules:
-----------------------------------------
Module				| Purpose
--------------------------------| -------
**Component.CastlingMove**	| Forked from module **Component.Move**.
**Data.Enum**			| Currently single-function.
**Data.Foldable**		| Currently single-function.
**Property.FixedMembership**	| Defines a class to which sum-types can conform.
**StateProperty.Censor**	| Relocated from directory **State/**.
**StateProperty.Mutator**	| defines a class to express the dual implementations within **State.Board**.
**StateProperty.Seeker**	| defines a class to express the dual implementations within **State.Board**.
**Text.Case**			| Forked from **Text.ShowList** to contain case-related operations.
**Text.Prefix**			| Forked from **Text.ShowList** to define the constant prefixes of log-messages.
-----------------------------------------

### Testing:
* Split **src-test/Main.hs** into **src-test/HUnit.hs** & **src-test/QuickCheck.hs**, each referenced independently from the cabal file.
* Added an executable **duel** (to coordinate a battle between two independently configured instances of **bishbosh**) & a corresponding section-1 man-page.
* Validated the list of ranks supplied to construct either **Attribute.RankValues.RankValues** or **Input.PieceSquareTable.PieceSquareTable**.

### Refactoring:
* Flattened the nested array **Component.Zobrist.getRandomByCoordinatesByRankByLogicalColour**, by means of a composite index.
* Reimplemented function **Cartesian.Coordinates.getLogicalColourOfSquare**.
* Reimplemented function **Cartesian.Coordinates.interpolationsByDestinationBySource** in terms of function **Cartesian.Coordinates.extrapolationsByDirectionByCoordinates**.
* Used the **LambdaCase** language-extension.

## 0.1.1.0
### New Modules:
---------------------------------------------------------
New Module				| Purpose
----------------------------------------| -------
**BishBosh.Time.StopWatch**		| Replaces module **BishBosh.Data.Time** to encapsulate interaction with module **Data.Time.Clock**.

src-exe/Duel.hs
{-# LANGUAGE LambdaCase #-}
{-
	Copyright (C) 2021 Dr. Alistair Ward

	This file is part of Duel.

	Duel is free software: you can redistribute it and/or modify
	it under the terms of the GNU General Public License as published by
	the Free Software Foundation, either version 3 of the License, or
	(at your option) any later version.

	Duel is distributed in the hope that it will be useful,
	but WITHOUT ANY WARRANTY; without even the implied warranty of
	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
	GNU General Public License for more details.

	You should have received a copy of the GNU General Public License
	along with Duel.  If not, see <http://www.gnu.org/licenses/>.
-}
{- |
 [@AUTHOR@]	Dr. Alistair Ward

 [@DESCRIPTION@]	Coordinates a duel between two independently configured instances of 'bishbosh'; which can be used to evaluate relative performance.

-}

module Main(main)	where

import			Control.Arrow((&&&))
import qualified	BishBosh.Data.Exception				as Data.Exception
import qualified	BishBosh.Input.CategorisedCommandLineOptions	as Input.CategorisedCommandLineOptions
import qualified	BishBosh.Input.CommandLineOption		as Input.CommandLineOption
import qualified	BishBosh.Input.Verbosity			as Input.Verbosity
import qualified	BishBosh.Text.Case				as Text.Case
import qualified	Control.Exception
import qualified	Data.Default
import qualified	Data.List
import qualified	Data.Version
import qualified	Duel.Data.Options				as Data.Options
import qualified	Duel.Process.Intermediary			as Process.Intermediary
import qualified	Paths_bishbosh					as Paths	-- Either local stub, or package-instance auto-generated by 'Setup build'.
import qualified	System.Console.GetOpt				as G
import qualified	System.Environment
import qualified	System.Exit
import qualified	System.Info

-- | The application's entry-point.
main :: IO ()
main	= do
	progName	<- System.Environment.getProgName

	let
		optDescrList :: [
			G.OptDescr (
				Maybe (Input.CommandLineOption.Flag, Maybe String {-optional value-}),	-- Record command-line specifications.
				Input.CommandLineOption.CommandLineOption Data.Options.Options		-- Defines the action to take on receipt of a command-line option.
			) -- Pair.
		 ]
		optDescrList	= [
			G.Option "?"	["help"] (
				G.NoArg (
					Nothing {-N/A-},
					Input.CommandLineOption.mkIOAction printHelp
				)
			) "print this help, & then exit.",
			G.Option "v"	["version"] (
				G.NoArg (
					Nothing {-N/A-},
					Input.CommandLineOption.mkIOAction printVersion
				)
			) "print version-information, & then exit.",
			G.Option ""	[Input.Verbosity.tag] (
				G.ReqArg (
					Just . (,) Input.Verbosity.tag . Just &&& Input.CommandLineOption.mkOptionsMutator . Data.Options.setVerbosity . Input.CommandLineOption.readArg
				) . mkTypeSpecification $ Text.Case.toUpperInitial Input.Verbosity.tag
			) . showString "define the log-level; default '" $ shows (Data.Default.def :: Input.Verbosity.Verbosity) "'.",
			let
				flag	= "nDecimalDigits"
			in G.Option "d"	[flag] (
				G.ReqArg (
					Just . (,) flag . Just &&& Input.CommandLineOption.mkOptionsMutator . Data.Options.setNDecimalDigits . Input.CommandLineOption.readBoundedIntegral
				) $ mkTypeSpecification "Int"
			) . showString "define the precision with which to display floating-point numbers; default " $ shows (Data.Options.getNDecimalDigits Data.Default.def) ".",
			let
				flag	= "nGames"
			in G.Option "n"	[flag] (
				G.ReqArg (
					Just . (,) flag . Just &&& Input.CommandLineOption.mkOptionsMutator . Data.Options.setNGames . Input.CommandLineOption.readBoundedIntegral
				) $ mkTypeSpecification "Int"
			) . showString "define the number of games to play; default " $ shows (Data.Options.getNGames Data.Default.def) ".",
			let
				flag	= "readTimeout"
			in G.Option "t"	[flag] (
				G.ReqArg (
					Just . (,) flag . Just &&& Input.CommandLineOption.mkOptionsMutator . Data.Options.setReadTimeout . Input.CommandLineOption.readBoundedIntegral
				) $ mkTypeSpecification "Seconds"
			) . showString "define the read-timeout; default " $ shows (Data.Options.getReadTimeout Data.Default.def) " s.",
			let
				flag	= "verifyConfiguration"
			in G.Option ""	[flag] (
				G.NoArg (
					Nothing {-N/A-},
					Input.CommandLineOption.mkOptionsMutator $ Data.Options.setVerifyConfiguration True
				)
			) "verify the mutual compatibility of the two configuration-files before forwarding each to a forked instance of 'bishbosh'.",
			let
				flag	= "appendInputConfigFilePath"
			in G.Option "i"	[flag] (
				G.ReqArg (
					Just . (,) flag . Just &&& Input.CommandLineOption.mkOptionsMutator . Data.Options.appendInputConfigFilePath
				) $ mkTypeSpecification "File-path"
			) "define the path to an XML-file, to forward to a forked instance of 'bishbosh'; specify White's first, then Black's."
		 ] where
			author :: String
			author	= "Dr. Alistair Ward"

			printHelp, printVersion :: Input.CategorisedCommandLineOptions.IOAction
			printHelp	= putStrLn . showString "Usage: " $ G.usageInfo progName optDescrList
			printVersion	= putStrLn . showString progName . showChar '-' . showsVersion Paths.version . showString "\n\nCompiled by " . showString System.Info.compilerName . showChar '-' . showsVersion System.Info.compilerVersion . showString ".\n\nCopyright (C) 2021 " . showString author . showString ".\nThis program comes with ABSOLUTELY NO WARRANTY.\nThis is free software, and you are welcome to redistribute it under certain conditions.\n\nWritten by " $ showString author "." where
				showsVersion :: Data.Version.Version -> ShowS
				showsVersion	= foldr (.) id . Data.List.intersperse (showChar '.') . map shows . Data.Version.versionBranch

			mkTypeSpecification :: ShowS
			mkTypeSpecification s	= showChar '<' $ showString s ">"

	System.Environment.getArgs >>= (
		\case
			(commandLineOptions, [{-non-options-}], [{-errors-}])
				| not $ null ioActions	-> sequence_ ioActions >> System.Exit.exitSuccess
				| otherwise		-> Process.Intermediary.initialise . foldr ($) Data.Default.def $ Input.CategorisedCommandLineOptions.getOptionsMutators categorisedCommandLineOptions	-- Sequentially mutate the configuration, using each of the options-mutators specified on the command-line.
				where
					categorisedCommandLineOptions		= Input.CommandLineOption.categorise $ map snd commandLineOptions
					ioActions				= Input.CategorisedCommandLineOptions.getIOActions categorisedCommandLineOptions
			(_, nonOptions, [])	-> Control.Exception.throwIO . Data.Exception.mkRedundantData . showString "unexpected command-line arguments; " $ shows nonOptions "."
			(_, _, errors)		-> Control.Exception.throwIO . Data.Exception.mkParseFailure $ concat errors
	 ) . G.getOpt G.RequireOrder optDescrList

